02-household-queries/dspy_engine.py

```diff
@@ -137,9 +177,199 @@ def create_retriever_model():
+        return dspy.OllamaLocal(model=llm_name, temperature=0.1)
+    elif llm_name in [
+        "gpt-3.5-turbo",
+        "gpt-3.5-turbo-instruct",
+        "gpt-4",
+        "gpt-4-turbo",
+        "gpt-4-turbo-instruct",
+    ]:
+        return dspy.OpenAI(model=llm_name, temperature=0.1)
+
+
+def main_baseline(query):
+    retrieve_k = int(os.environ.get("RETRIEVE_K", "2"))
+
+    dspy.settings.configure(lm=create_llm_model(), rm=create_retriever_model())
+
+    # run_basic_predictor(query)
+    # run_cot_predictor(query)
+    # run_retrieval(query, retrieve_k)
+    run_rag(query, retrieve_k)
+
+
+def main_train(qa_pairs):
+    training = [qa.with_inputs("question") for qa in qa_pairs]
+    print_training_data(training)
+
+    retrieve_k = int(os.environ.get("RETRIEVE_K", "10"))
+    print("Using retrieve_k =", retrieve_k)
+    module: dspy.Module = RAG(retrieve_k)
+
+    filename = input(
+        "Specify compiled_module file to load optimized module or press Enter to start from scratch: "
+    )
+    if filename:
+        module.load(filename)
+        compiled_module = module
+
+    # Options: validate_answer, validate_context_and_answer
+    eval_metric = validate_context_and_answer
+
+    if not filename:
+        optimizer_choice = input(f"""Which optimizer?
+        1. COPRO(metric={eval_metric.__name__})
+        2. LabeledFewShot(3)
+        3. BootstrapFewShot(metric={eval_metric.__name__})
+    """)
+        eval_kwargs = None
+        if optimizer_choice == "1":
+            # Can specify prompt_model to use different LLM for prompt candidate generation
+            # By default, it uses the same LLM as dspy.settings.lm
+            optimizer = COPRO(metric=eval_metric, verbose=True)
+            eval_kwargs = dict(num_threads=64, display_progress=True, display_table=0)
+        elif optimizer_choice == "2":
+            optimizer = LabeledFewShot(3)
+        elif optimizer_choice == "3":
+            optimizer = BootstrapFewShot(metric=eval_metric)
+
+    llm_choice = input("""Which LLM?
+        openhermes (default), llama2, mistral,
+        gpt-3.5-turbo, gpt-3.5-turbo-instruct, gpt-4, gpt-4-turbo, gpt-4-turbo-instruct
+    """)
+    if not llm_choice:
+        llm_choice = "openhermes"
+    dspy.settings.configure(
+        lm=create_llm_model(llm_choice), rm=create_retriever_model()
+    )
+
+    if not filename:
+        file_suffix = f"optimizer{optimizer_choice}_{retrieve_k}_{llm_choice}_{time.strftime('%Y-%m-%d-%H%M%S')}"
+        compiled_module = optimize_module(
+            optimizer, module, training, eval_kwargs, file_suffix
+        )
+    print_optimization(compiled_module)
+
+    evaluate_module(compiled_module, eval_metric, training)
+
+
+# Custom validator for training
+# Built-in alternatives:
+# - dspy.evaluate.metrics.answer_exact_match
+# - dspy.evaluate.metrics.answer_passage_match
+def validate_answer(example, pred, trace=None):
+    return pred.answer.lower().startswith(example.answer.lower())
```

Comment:
FWIW I've been doing something like `pred.answer.lower() in example.answer.lower()` because I've noticed that even when they're asked to just return a single word, sometimes LLMs will start include more boilerplate (e.g., in the evaluation code, sometimes I get `Correct`, `correct`, `"Correct"`, `The answer is correct.`, etc.)

Reply:
👍 I considered that, however I didn't want it to find 'no' somewhere else like in 'know'. In my case, the responses almost always have the answer as the first word.

Reply:
Ah good call. Yeah, I had a similar issue with my original implementation checking for `correct` which is of course present in... `incorrect`